### PR TITLE
Guidance about serialization

### PR DESCRIPTION
**Pull request by atanassov**
Adding a general guidance about serialization for issue #284

***
<a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/pull/367.html" title="Last updated on Apr 20, 2023, 1:28 AM UTC (f257a08)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/367/e81cae6...f257a08.html" title="Last updated on Apr 20, 2023, 1:28 AM UTC (f257a08)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -2680,6 +2680,22 @@ which meant converting the names to camel-cased version.
 One such example is the <a href="https://w3c.github.io/image-resource/">image resource</a>.
 For this reason, if a key can clearly be expressed as a single word, that is recommended.
 
+<h3 id="serialization">Serialization</h3>
+Adding or extending features that involve a parser or a serializer should consider their affect to 
+serialization. The following are constituencies of serialization results that must be considered:
+
+ - Users - often reading result of serialization to validate expectations or learn
+ - Tools - use these results to enrich dev experience, provide auditing, automation, testing etc.
```
Comment (atanassov) [inline index.bs:2688]:
```suggestion
 - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
```

File: index.bs [hunk 2]
```diff
@@ -2680,6 +2680,22 @@ which meant converting the names to camel-cased version.
 One such example is the <a href="https://w3c.github.io/image-resource/">image resource</a>.
 For this reason, if a key can clearly be expressed as a single word, that is recommended.
 
+<h3 id="serialization">Serialization</h3>
+Adding or extending features that involve a parser or a serializer should consider their affect to
```
Comment (atanassov) [inline index.bs:2684]:
```suggestion
Adding or extending features that involve a parser or a serializer should consider their effect on 
```

File: index.bs [hunk 3]
```diff
@@ -2680,6 +2680,22 @@ which meant converting the names to camel-cased version.
...
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
+
+There are also specific expectations that can be language specific - significance of empty spaces, 
+numeric precision, structure of the format (ex. CSS grid structure) etc. In such cases the expected 
+effect to serialization results:
+
+ - Should be intuitive (this is rather weakly defined expectation but often cited in discussions)
+ - Should not add to error accumulation - taking the serialized output of an API and feeding it 
+ back to the same API in a loop should be predictable and ideally error free (at least for simple API)
```
Comment (atanassov) [inline index.bs:2697]:
```suggestion
 back to the same API in a loop should result in the same internal state, and ideally error free (at least for simple API)
```

File: index.bs [hunk 4]
```diff
@@ -2680,6 +2680,22 @@ which meant converting the names to camel-cased version.
...
 
+<h3 id="serialization">Serialization</h3>
+Adding or extending features that involve a parser or a serializer should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
+
+ - Users - often reading result of serialization to validate expectations or learn
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
+
+There are also specific expectations that can be language specific - significance of empty spaces,
```
Comment (atanassov) [inline index.bs:2691]:
```suggestion
There are also expectations that can be language specific - significance of empty spaces, 
```

File: index.bs [hunk 5]
```diff
@@ -2680,6 +2680,22 @@ which meant converting the names to camel-cased version.
...
 
+<h3 id="serialization">Serialization</h3>
+Adding or extending features that involve a parser or a serializer should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
+
+ - Users - often reading result of serialization to validate expectations or learn
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
+
+There are also expectations that can be language specific - significance of empty spaces,
```
Comment (atanassov) [inline index.bs:2691]:
```suggestion
Consider expectations that can be language specific - significance of empty spaces, 
```

File: index.bs [hunk 6]
```diff
@@ -2680,6 +2680,22 @@ which meant converting the names to camel-cased version.
...
 
+<h3 id="serialization">Serialization</h3>
+Adding or extending features that involve a parser or a serializer should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
+
+ - Users - often reading result of serialization to validate expectations or learn
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
+
+Consider expectations that can be language specific - significance of empty spaces,
```
Comment (atanassov) [inline index.bs:2691]:
```suggestion
Consider language specific expectations - significance of empty spaces, 
```

File: index.bs [hunk 7]
```diff
@@ -2680,6 +2680,22 @@ which meant converting the names to camel-cased version.
...
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
+
+Consider language specific expectations - significance of empty spaces, 
+numeric precision, structure of the format (ex. CSS grid structure) etc. In such cases the expected 
+effect to serialization results:
+
+ - Should be intuitive (this is rather weakly defined expectation but often cited in discussions)
+ - Should not add to error accumulation - taking the serialized output of an API and feeding it 
+ back to the same API in a loop should result in the same internal state and ideally error free (at least for simple API)
```
Comment (annevk) [inline index.bs:2697]: It should also result in the same serialization. I.e., you want idempotence for serialize(parse(input)).

File: index.bs [hunk 8]
```diff
@@ -2680,6 +2680,22 @@ which meant converting the names to camel-cased version.
 One such example is the <a href="https://w3c.github.io/image-resource/">image resource</a>.
 For this reason, if a key can clearly be expressed as a single word, that is recommended.
 
+<h3 id="serialization">Serialization</h3>
```
Comment (LeaVerou) [inline index.bs:2683]: Title should be an imperative summary of guidance, [per newer meta - design principle](https://github.com/w3ctag/design-principles/pull/371)

File: index.bs [hunk 9]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
...
+Adding or extending features that involve a parser or a serializer should consider their effect on
```
Comment (torgo) [inline index.bs:2806]:
```suggestion
Adding or extending features that involve a serializer should consider their effect on 
```

File: index.bs [hunk 10]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
 One such example is the <a href="https://w3c.github.io/image-resource/">image resource</a>.
 For this reason, if a key can clearly be expressed as a single word, that is recommended.
 
+<h3 id="serialization">Serialization</h3>
+Adding or extending features that involve a parser or a serializer should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
```
Comment (torgo) [inline index.bs:2807]:
```suggestion
Adding or extending features that involve a serializer should consider their effect on 
consumers. The following are constituencies of serialization results that must be considered:
```

File: index.bs (hunk 8 above)
Comment (torgo) [reply index.bs:2805]:
```suggestion
<h3 id="serialization">Consider Consumers when Serializing</h3>
```

**Review by torgo — APPROVED** [review]
lgtm

**Comment by torgo** [comment]
@atanassov we made some changes and we think this is ready to merge. Look good to you?

File: index.bs [hunk 11]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
...
+ - Users - often reading result of serialization to validate expectations or learn
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
+
+Consider language specific expectations - significance of empty spaces, 
+numeric precision, structure of the format (ex. CSS grid structure) etc. In such cases the expected 
+effect to serialization results:
+
+ - Should be intuitive (this is rather weakly defined expectation but often cited in discussions)
+ - Should not add to error accumulation - taking the serialized output of an API and feeding it
```
Comment (hober) [inline index.bs:2818]:
```suggestion
  - [Idempotence](https://en.wikipedia.org/wiki/Idempotence) - the result of serializing the output of a parser should be something that, when parsed and serialized, produces itself. 
  - Should not add to error accumulation - taking the serialized output of an API and feeding it 
```

Comment (hober) [reply index.bs:2818]: Per @annevk's comment.

File: index.bs (hunk 10 above)
Comment (hober) [reply index.bs:2807]: I think the original text is clearer,

**Review by hober — CHANGES_REQUESTED** [review]

File: index.bs [hunk 12]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
 One such example is the <a href="https://w3c.github.io/image-resource/">image resource</a>.
 For this reason, if a key can clearly be expressed as a single word, that is recommended.
 
+<h3 id="serialization">Consider consumers when serializing</h3>
+Adding or extending features that involve a parser or a serializer should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
+
+ - Users - often reading result of serialization to validate expectations or learn
```
Comment (hober) [inline index.bs:2809]:
```suggestion
 - Users - because the result of serialization may be presented to the end user
```

File: index.bs [hunk 13]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
 One such example is the <a href="https://w3c.github.io/image-resource/">image resource</a>.
 For this reason, if a key can clearly be expressed as a single word, that is recommended.
 
+<h3 id="serialization">Consider consumers when serializing</h3>
+Adding or extending features that involve a parser or a serializer should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
```
Comment (torgo) [inline index.bs:2807]:
```suggestion
When adding or extending features that involve a parser or a serializer you should consider their effect on 
serialization. The following are constituencies of serialization results that must be considered:
```

File: index.bs [hunk 14]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
...
+
+ - Users - often reading result of serialization to validate expectations or learn
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
```
Comment (hober) [inline index.bs:2810]:
```suggestion
 - Tools - which may rely on serialization in many different ways
```

File: index.bs [hunk 15]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
...
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
```
Comment (hober) [inline index.bs:2811]:
```suggestion
 - Web APIs - because serializations may be  passed into other APIs of the web platform
```

File: index.bs [hunk 16]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
...
+
+Consider language specific expectations - significance of empty spaces,
```
Comment (hober) [inline index.bs:2813]:
```suggestion
Consider language specific expectations - for instance, in some languages, the presence or absence of whitespace may be significant. 
```

File: index.bs [hunk 17]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
...
+<h3 id="serialization">Consider consumers when serializing</h3>
+Adding or extending features that involve a parser or a serializer should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
+
+ - Users - often reading result of serialization to validate expectations or learn
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
+
+Consider language specific expectations - significance of empty spaces, 
+numeric precision, structure of the format (ex. CSS grid structure) etc. In such cases the expected
```
Comment (hober) [inline index.bs:2814]:
```suggestion
Languages may differ in the precision of floating point numbers they accept.
```

File: index.bs [hunk 18]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
...
+Adding or extending features that involve a parser or a serializer should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
+
+ - Users - often reading result of serialization to validate expectations or learn
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
+ - Web platform API - serialized results passed as input to other (or the same) API of the platform
+
+Consider language specific expectations - significance of empty spaces, 
+numeric precision, structure of the format (ex. CSS grid structure) etc. In such cases the expected 
+effect to serialization results:
```
Comment (hober) [inline index.bs:2815]:
```suggestion
The results of serialization:
```

File: index.bs [hunk 19]
```diff
@@ -2802,6 +2802,22 @@ which meant converting the names to camel-cased version.
...
+
+ - Should be intuitive (this is rather weakly defined expectation but often cited in discussions)
```
Comment (hober) [inline index.bs:2817]:
```suggestion
 - Should match developer expectations
```

**Review by hober — COMMENTED** [review]
Here's a general pass on the text for readability.

File: index.bs (hunk 14 above)
Comment (torgo) [reply index.bs:2810]: feels like a couple of examples would be good

File: index.bs (hunk 19 above)
Comment (torgo) [reply index.bs:2817]: maybe add a parenthetical example

File: index.bs (hunk 7 above)
Comment (hober) [reply index.bs]: Added a bullet about idemptence. Thanks, @annevk!

File: index.bs [hunk 20]
```diff
@@ -2802,6 +2802,23 @@ which meant converting the names to camel-cased version.
 One such example is the <a href="https://w3c.github.io/image-resource/">image resource</a>.
 For this reason, if a key can clearly be expressed as a single word, that is recommended.
 
+<h3 id="serialization">Consider consumers when serializing</h3>
+When adding or extending features that involve a parser or a serializer you should consider their effect on 
+serialization. The following are constituencies of serialization results that must be considered:
+
+ - Users - because the result of serialization may be presented to the end user
+ - Tools - use these results to enrich author experience, provide auditing, automation, testing etc.
```
Comment (hober) [inline index.bs:2810]:
```suggestion
 - Tools - which may rely on the output of serialization in a number of ways (for instance, to detect if the parser need to correct for errors in the input)
```

File: index.bs [hunk 21]
```diff
@@ -2802,6 +2802,23 @@ which meant converting the names to camel-cased version.
...
+ - Web APIs - because serializations may be  passed into other APIs of the web platform
+
+Consider language specific expectations - for instance, in some languages, the presence or absence of whitespace may be significant. 
+Languages may differ in the precision of floating point numbers they accept.
+The results of serialization:
+
+ - Should be intuitive (this is rather weakly defined expectation but often cited in discussions)
```
Comment (hober) [inline index.bs:2817]:
```suggestion
 - Should match developer expectations (for instance, serializing CSS properties should not result in output that is very different from what a CSS author would have written)
```

**Review by hober — APPROVED** [review]

**Review by torgo — APPROVED** [review]
lgtm

**Pull request merged**